src/compiler/checker.ts

```diff
@@ -67,6 +67,7 @@ namespace ts {
         let enumCount = 0;
         let instantiationDepth = 0;
         let constraintDepth = 0;
+        let simplificationDepth = 0;

```

Comment:
What test necessitates tracking this?

Reply:
Like, why aren't we simplifying to an arbitrary number of levels deep? `getSimplifiedType` already has a reentrancy guard in place because conditional types caused it to be reentrant.

Reply:
tests/cases/compiler/infiniteConstraints.ts

Reply:
Looking into it, deleting the `if (objectType.flags & TypeFlags.TypeParameter) {` case within `getSimplifiedIndexedAccessType` fixes the bad generativeness of `infiniteConstraints.ts` (since it's only by diving into a constraint that you could possibly be manufacturing new types during this process). I don't think it should be necessary for anything (why is something that "simplifies" looking at constraints?) and is the only reason that generative-ness can occur. From what I can tell, it's a vestige of when we were missing some assignability rules for indexed accesses, and removing it + adding the `return mapType(instantiateType(getTemplateTypeFromMappedType(objectType), templateMapper), getSimplifiedType);` line in `substituteIndexedMappedType` results in one new error in `keyofAndIndexedAccess.ts` which, looking at it is _correct_.

Reply:
And I do so love deleting code.

Reply:
I think you're right. It does mean that the repro from #13285 will break, but I think that is correct given how it is currently written. It may in fact not be possible to write that example without any type annotations or assertions.

Reply:
Aye - the input map type `{[oldId: string]: string}` is too general.

Reply:
I've updated the PR, including the closest we can get in the repro from #13285. Unfortunately we need an explicit type annotation because we can't narrow generic types that have `undefined` in their constraint. But that's a known issue.